### PR TITLE
PostgreSQL pointerdb storage

### PR DESCRIPTION
**Pull request by thepaul**
The client_test here requires a postgresql instance to exist already, with a suitable user and database. Still figuring out how to fix that up, but at least the storage testsuite passes and captplanet works with it and everything. Would like to get some review feedback on this if possible before I go too far down a wrong path.

### CODE REVIEW

**Comment by egonelbre** [comment]
For things like integration tests I tend to add a flag for it, like (grabbed from some of my code):

```
var testPostgres = flag.String("postgres-test-db", "", "postgres test database")

func newTestPostgres(t *testing.T) (store *PostgresDB, cleanup func()) {
	if *testPostgres == "" {
		t.Skip(`postgres flag missing, example:` + "\n" + `-postgres-test-db="user=pointerdb password=q1w2e3r4 dbname=pointerdb"`)
	}

	pgdb, err := NewPostgresDB(*testPostgres)
	if err != nil {
		t.Fatalf("init: %v", err)
	}

	return pgdb, func() {
		if err := pgdb.DestroyDatabase(); err != nil {
			t.Fatal(err)
		}

		if err := pgdb.Close(); err != nil {
			t.Fatal(err)
		}
	}
}

func TestCRUD(t *testing.T) {
    db, cleanup := newTestPostgres(t)
    defer cleanup()

    ...
}
```

File: pkg/pointerdb/config.go [hunk 1]
```diff
@@ -30,25 +32,32 @@ type Config struct {
 	Overlay              bool   `default:"false" help:"toggle flag if overlay is enabled"`
 }
 
-// Run implements the provider.Responsibility interface
-func (c Config) Run(ctx context.Context, server *provider.Provider) error {
-	dburl, err := utils.ParseURL(c.DatabaseURL)
+func setupKVDriver(dbURLString string) (db storage.KeyValueStore, err error) {
```
Comment (egonelbre) [inline pkg/pointerdb/config.go]: `newKeyValueStore`

File: storage/postgreskv/client.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		start = opi.opts.First
+	}
+	var query string
+	if !opi.opts.Recurse {
+		if opi.opts.Reverse {
+			query = "SELECT p, m FROM list_directory_reverse($1::BYTEA, $2::BYTEA, $3::BYTEA, $4) ld(p, m)"
+		} else {
+			query = "SELECT p, m FROM list_directory($1::BYTEA, $2::BYTEA, $3::BYTEA, $4) ld(p, m)"
+		}
+	} else {
```
Comment (egonelbre) [inline storage/postgreskv/client.go]: The list_directory functions seem very complicated, it seems like there should be an easier way to handle it. I'm probably missing something essential, but anyways... (ignore all the postgres code bugs and +-1 errors, pseudo-code to show the ideas)

For recursive, this should LIKE queries should go directly against B-Tree so it can be implemented as:

```
SELECT fullpath, value
FROM path
WHERE fullpath LIKE {prefix}%
ORDER BY fullpath
WHERE {start} <= fullpath and fullpath <= {end} 
LIMIT $4
```

For non-recursive I'm less sure about the performance, could go either way... but the implementation could be much more concise.

```
SELECT 
    first(fullpath),
    localPath(fullpath, {prefix}, {delimiter}) as localPath,
    count(*),
    first(value)
FROM path
WHERE fullpath like {prefix}%
ORDER BY fullpath
GROUP BY localPath
WHERE ${start} <= fullpath and fullpath <= {end} 
LIMIT $4

function localPath(fullpath, prefix, delimiter) {
    // I think you have it as truncate_path, although not completely the same
    localFullPath = substring(fullpath from length(prefix))
    p = position(delimiter in localFullPath)
    if p < 0 return localFullPath
    return substring(localFullPath for p)
}
```

Comment (thepaul) [reply storage/postgreskv/client.go]: For recursive queries, that's essentially what is being used already. That query is assembled in orderedPostgresIterator.nextQuery().

For non-recursive, I made a medium-large dataset of paths (all the filename paths on my workstation) and tested the performance of a variety of solutions. The performance from doing any string operations on every row, as you suggest, is prohibitively slow when there are a lot of subcomponents to skip over. Using DISTINCT on a function like this localPath() is even worse. Putting all elements into an adjacency-model tree gets barely-acceptable performance for the various types of read queries that are necessary, but is slow for inserts or moves and somewhat costly in storage space. Building trees with the nested-sets or nested-intervals models is not a great fit, because of the structure of the data. Storing materialized paths as postgresql arrays works very poorly. Postgres's own ltree type (from contrib) might work well, but it imposes limitations on the size of paths and path components that I don't think are tenable here. It would be very nice if pg could apply partitioning or DISTINCT on a LIKE prefix expression, but according to what I've been able to find, it can't. If you have conflicting data, I can investigate further.

In short, this recursive CTE is the only approach I can find which yields efficient lookups _and_ efficient inserts. As a bonus, it's nearly optimal in terms of storage space as well. The downside is the very ugly and hard-to-understand query structure.

File: storage/postgreskv/client.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		if opi.opts.Reverse {
+			startCmp = "<="
+			orderDir = " DESC"
+		}
+		query = fmt.Sprintf(`
+			SELECT fullpath, metadata
+			  FROM pathdata
+			 WHERE bucket = $1::BYTEA
+			   AND ($2::BYTEA = ''::BYTEA OR fullpath >= $2::BYTEA)
+			   AND ($2::BYTEA = ''::BYTEA OR fullpath < bytea_increment($2::BYTEA))
```
[the remark this answers is not in the document]
Comment (thepaul) [reply storage/postgreskv/client.go]: bytea_increment would still need to be defined in the SQL, though, because it's used in the recursive CTE. i chose to make use of the one in PG here to avoid implementing it twice and running the risk of the two diverging.

File: storage/postgreskv/client.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+	`
+	rows, err := client.pgConn.Query(q, []byte(bucket), pq.ByteaArray(keys.ByteSlices()))
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	values := make([]storage.Value, 0, len(keys))
+	for rows.Next() {
+		var value []byte
+		if err := rows.Scan(&value); err != nil {
+			_ = rows.Close()
```
[the remark this answers is not in the document]
Comment (thepaul) [reply storage/postgreskv/client.go]: In this case, we already have an error from rows.Scan. Is it still worthwhile to collect an error from rows.Close() and combine the two?

**Comment by thepaul** [comment]
Thanks. I'll add a flag as suggested and resolve the other items.

File: storage/postgreskv/client.go (hunk 2 above)
Comment (egonelbre) [reply storage/postgreskv/client.go]: Ah, awesome. Can you add a relevant benchmark for the listing?

File: storage/postgreskv/client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (egonelbre) [reply storage/postgreskv/client.go]: Yes, I think so. We have `utils.CombineErrors`. Here, I mainly just don't want to have a precedent of not checking errors. I'm aware that they are fine, but other people use other code as a reference so, it's probably better to check them.

File: storage/postgreskv/client.go (hunk 2 above)
Comment (thepaul) [reply storage/postgreskv/client.go]: Ok. I'll put something together.

**Comment by thepaul** [comment]
So I gather that travis doesn't use the docker-compose thing. I'll dig into how to make a pg server available during test.

File: .travis.yml [hunk 5]
```diff
@@ -33,7 +35,7 @@ matrix:
         - popd
         - go install -race ./...
       script:
-        - go test -race -cover -coverprofile=.coverprofile ./...
+        - POSTGRESKV_TEST='postgres://postgres@localhost/pointerdb?sslmode=disable' go test -race -cover -coverprofile=.coverprofile ./...
```
Comment (egonelbre) [inline .travis.yml]: Note you can use:
```
- env:
    - MODE=tests
    - STORJ_POSTGRES_TEST='....'
  ...
```

File: storage/postgreskv/alternateclient_test.go [hunk 6]
```diff
@@ -13,9 +15,21 @@ import (
 	"storj.io/storj/storage/testsuite"
 )
 
+var (
+	doAltTests = flag.Bool("test-postgreskv-alt", false, "Run the KeyValueStore tests against the alternate PG implementation")
```
Comment (egonelbre) [inline storage/postgreskv/alternateclient_test.go]:
```
postgresTestAddr = flag.String("postgres-addr", os.Getenv("STORJ_POSTGRES_TEST"), "Run the KeyValueStore tests against the alternate PG implementation")
```

Then in tests you can check for `if *postgresTestAddr == "" { t.Skip ...`

File: .travis.yml (hunk 5 above)
Comment (thepaul) [reply .travis.yml]: Ok, that would be a lot cleaner. I'm finding travis's yml schema to be quite confusing.

File: storage/postgreskv/alternateclient_test.go (hunk 6 above)
Comment (thepaul) [reply storage/postgreskv/alternateclient_test.go]: You mean, to allow for supplying a different postgresql database when running the alt-tests than the one used for the main postgres tests? That doesn't seem like a very pressing need.

Comment (egonelbre) [reply storage/postgreskv/alternateclient_test.go]: I mean is there a reason you are skipping the alternate tests? Although now thinking about, yeah.. it might be fine.

But, yeah, just use the same database. But, for the other flag you can use `os.Getenv("STORJ_POSTGRES_TEST")` as the default parameter to make the code shorter.

Comment (thepaul) [reply storage/postgreskv/alternateclient_test.go]: Yeah, I don't expect the "alternate client" to be around for long. I really just want it to be in the history in case it's needed for future reference. And I don't want to clutter up real, regular test results with tests of that throwaway code.

File: storage/postgreskv/client.go (hunk 2 above)
Comment (thepaul) [reply storage/postgreskv/client.go]: Some raw output from running the long benchmarks against pg, an alternate version of the pg client that doesn't use a recursive CTE, and against BoltDB:

```
~/work/storj% STORJ_POSTGRESKV_TEST='postgres://pointerdb:pg-secret-pass@localhost/pointerdb?sslmode=disable' go test -v -run xxx -bench BenchmarkSuiteLong/. ./storage/postgreskv/... -test-postgreskv-alt -test-bench-long
goos: darwin
goarch: amd64
pkg: storj.io/storj/storage/postgreskv
BenchmarkSuiteLongAlt/DeepRecursive-8         	       1	3972012305 ns/op	  12.90 MB/s
--- BENCH: BenchmarkSuiteLongAlt/DeepRecursive-8
    long_bench.go:251: Performing bulk import...
    long_bench.go:211: importing took 40.173655483s
BenchmarkSuiteLongAlt/DeepRecursiveReverse-8  	       1	3808230282 ns/op	  11.97 MB/s
BenchmarkSuiteLongAlt/DeepNonRecursive-8      	       1	9526808366 ns/op	   0.00 MB/s
BenchmarkSuiteLongAlt/DeepNonRecursiveReverse-8         	       1	9312339647 ns/op	   0.00 MB/s
BenchmarkSuiteLongAlt/ShallowRecursive-8                	       3	 460747219 ns/op	   2.68 MB/s
BenchmarkSuiteLongAlt/ShallowRecursiveReverse-8         	       3	 454055201 ns/op	   2.72 MB/s
BenchmarkSuiteLongAlt/ShallowNonRecursive-8             	      10	 151278584 ns/op	   8.17 MB/s
BenchmarkSuiteLongAlt/ShallowNonRecursiveReverse-8      	      10	 154597266 ns/op	   7.99 MB/s
BenchmarkSuiteLongAlt/TopRecursiveLimit-8               	       1	1114497280 ns/op	  91.73 MB/s
BenchmarkSuiteLongAlt/TopRecursiveLimitReverse-8        	       1	1190270426 ns/op	 106.29 MB/s
BenchmarkSuiteLongAlt/TopRecursiveStartAt-8             	       1	1100325170 ns/op	  97.58 MB/s
BenchmarkSuiteLongAlt/TopRecursiveStartAtReverse-8      	       2	 657334097 ns/op	  91.22 MB/s
BenchmarkSuiteLongAlt/TopNonRecursive-8                 	       1	38930303580 ns/op	   0.00 MB/s
BenchmarkSuiteLongAlt/TopNonRecursiveReverse-8          	       1	38593242354 ns/op	   0.00 MB/s
--- BENCH: BenchmarkSuiteLongAlt
    long_bench.go:775: Performing bulk cleanup...
    long_bench.go:233: cleanup took 95.361205ms
BenchmarkSuiteLong/DeepRecursive-8                      	       1	3932688969 ns/op	  13.03 MB/s
--- BENCH: BenchmarkSuiteLong/DeepRecursive-8
    long_bench.go:251: Performing bulk import...
    long_bench.go:211: importing took 39.139060379s
BenchmarkSuiteLong/DeepRecursiveReverse-8               	       1	3800757332 ns/op	  11.99 MB/s
BenchmarkSuiteLong/DeepNonRecursive-8                   	       1	1986910406 ns/op	   0.00 MB/s
BenchmarkSuiteLong/DeepNonRecursiveReverse-8            	       1	1298669505 ns/op	   0.00 MB/s
BenchmarkSuiteLong/ShallowRecursive-8                   	       3	 515943335 ns/op	   2.40 MB/s
BenchmarkSuiteLong/ShallowRecursiveReverse-8            	       3	 473977898 ns/op	   2.61 MB/s
BenchmarkSuiteLong/ShallowNonRecursive-8                	       1	4243869744 ns/op	   0.29 MB/s
BenchmarkSuiteLong/ShallowNonRecursiveReverse-8         	       1	2639730536 ns/op	   0.47 MB/s
BenchmarkSuiteLong/TopRecursiveLimit-8                  	       1	1074987222 ns/op	  95.10 MB/s
BenchmarkSuiteLong/TopRecursiveLimitReverse-8           	       1	1182386950 ns/op	 107.00 MB/s
BenchmarkSuiteLong/TopRecursiveStartAt-8                	       1	1111324959 ns/op	  96.61 MB/s
BenchmarkSuiteLong/TopRecursiveStartAtReverse-8         	       2	 665344640 ns/op	  90.12 MB/s
BenchmarkSuiteLong/TopNonRecursive-8                    	       2	 738233795 ns/op	   0.00 MB/s
BenchmarkSuiteLong/TopNonRecursiveReverse-8             	       2	 734726361 ns/op	   0.00 MB/s
--- BENCH: BenchmarkSuiteLong
    long_bench.go:775: Performing bulk cleanup...
    long_bench.go:233: cleanup took 93.600617ms
PASS
ok  	storj.io/storj/storage/postgreskv	231.662s
?   	storj.io/storj/storage/postgreskv/schema	[no test files]
~/work/storj% go test -v -run xxx -bench BenchmarkSuiteLong/. ./storage/boltdb/... -test-bench-long
goos: darwin
goarch: amd64
pkg: storj.io/storj/storage/boltdb
BenchmarkSuiteLong/DeepRecursive-8         	      10	 131461868 ns/op	 389.89 MB/s
--- BENCH: BenchmarkSuiteLong/DeepRecursive-8
    long_bench.go:251: Performing bulk import...
    long_bench.go:211: importing took 2m6.711399139s
BenchmarkSuiteLong/DeepRecursiveReverse-8  	      10	 142506496 ns/op	 319.80 MB/s
BenchmarkSuiteLong/DeepNonRecursive-8      	   30000	     56642 ns/op	  74.38 MB/s
BenchmarkSuiteLong/DeepNonRecursiveReverse-8         	   30000	     54868 ns/op	  76.78 MB/s
BenchmarkSuiteLong/ShallowRecursive-8                	     300	   4765679 ns/op	 259.34 MB/s
BenchmarkSuiteLong/ShallowRecursiveReverse-8         	     300	   4923633 ns/op	 251.02 MB/s
BenchmarkSuiteLong/ShallowNonRecursive-8             	     300	   5716027 ns/op	 216.23 MB/s
BenchmarkSuiteLong/ShallowNonRecursiveReverse-8      	     200	   5973946 ns/op	 206.89 MB/s
BenchmarkSuiteLong/TopRecursiveLimit-8               	       5	 301675698 ns/op	 338.88 MB/s
BenchmarkSuiteLong/TopRecursiveLimitReverse-8        	       3	 355463711 ns/op	 355.92 MB/s
BenchmarkSuiteLong/TopRecursiveStartAt-8             	       5	 304097570 ns/op	 353.08 MB/s
BenchmarkSuiteLong/TopRecursiveStartAtReverse-8      	       5	 202944901 ns/op	 295.46 MB/s
BenchmarkSuiteLong/TopNonRecursive-8                 	  100000	     13864 ns/op	  17.24 MB/s
BenchmarkSuiteLong/TopNonRecursiveReverse-8          	  100000	     13232 ns/op	  18.06 MB/s
--- BENCH: BenchmarkSuiteLong
    long_bench.go:775: Performing bulk cleanup...
    long_bench.go:233: cleanup took 258.015µs
PASS
ok  	storj.io/storj/storage/boltdb	154.925s
```

The above numbers, graphed:

chart of storj long_bench times

BoltDB does extremely well for all the reading tasks, but quite poorly under heavy write load (and that is with syncs turned off). If I remember how lmdb/bolt are implemented, I'd expect bolt to start showing worse numbers in comparison to postgresql when there are more simultaneous queries. Or I might be totally wrong about that.

The alternate/straightforward PG approach does well, except in the case where there are lots of elements that need to be skipped over, such as when getting a listing of the toplevel directory. In that case, though, performance is utterly abysmal.

The recursive-CTE PG approach does more poorly when getting a listing of a directory with thousands of entries but little to no subdirectory items that need to be skipped over: a "shallow" directory. This is to be expected, as it requires extra calculation and new index lookups after every element it finds. But even in this worst case, performance is maaaaybe within the bounds of sanity.

Perhaps we could come up with a solution where both approaches are used according to some heuristics that try to predict which will do better. Or, if there are cpu cycles to spare, just run both queries each time and use whichever one returns first.

I'd like to implement some of the other solutions I mentioned, to the point that they pass tests, and get benchmark information from all of them, but I'm afraid I've probably spent more time on this than it deserves for now.

File: storage/testsuite/long_bench.go [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+	initialized := false
+
+	// create these anew with each invocation of this function; we only want these to run "once"
+	// per storage backend under test.
+	var initOnce sync.Once
+	var cleanupOnce sync.Once
+
+	doTest := func(name string, testFunc func(*testing.B, storage.KeyValueStore)) {
+		b.Run(name, func(bb *testing.B) {
+			initOnce.Do(func() { initStore(bb, store, &initialized) })
```
Comment (egonelbre) [inline storage/testsuite/long_bench.go]: You don't need to use sync.Once things. It's sufficient if you do in the beginning of function:

```
if !*disableInit { initStore(b, store) }
if !*disableCleanup { defer cleanupStore(b, store) }
```

File: storage/boltdb/client_test.go [hunk 8]
```diff
@@ -78,3 +80,53 @@ func TestSuiteShared(t *testing.T) {
...
+			return fmt.Errorf("Failed to insert data (%q, %q): %v", item.Key, item.Value, err)
+		}
+	}
+
+	return store.db.Sync()
+}
+
+func (store *boltLongBenchmarkStore) BulkDelete() error {
+	// just delete all the things!
+	return os.RemoveAll(store.dirPath)
```
[the remark this answers is not in the document]
Comment (thepaul) [reply storage/boltdb/client_test.go]: Doh, good point. It's been many years since I wrote anything targeting windows. I suppose this should be a no-op function, then, because everything will get cleaned up afterward in this boltdb case.

File: storage/testsuite/long_bench.go [hunk 9]
```diff
@@ -0,0 +1,801 @@
...
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/utils"
+	"storj.io/storj/storage"
+)
+
+const (
+	pathDataFileName = "paths.data.gz"
```
[the remark this answers is not in the document]
Comment (thepaul) [reply storage/testsuite/long_bench.go]: If we were encrypting entire paths and padding them out to a constant size, then yes, it would be indistinguishable from random data and that would work fine. However, we're only encrypting path components, while keeping the directory structure intact. The test data ought to have a natural hierarchy to it for the tests to be optimal. I don't know that anyone has done the research necessary to model a typical organic filesystem hierarchy with random data.

As for the large binary file, do we not have a way to ship the product without including source and test artifacts? Or do you mean the initial source download? I don't think I've ever worried much about a largish source tarball download.

File: storage/testsuite/long_bench.go (hunk 7 above)
Comment (thepaul) [reply storage/testsuite/long_bench.go]: I did it like this because I thought I had found a case where the import and cleanup were still being run without any tests, because of the filtering options I had applied, and I didn't want those to run unless something was actually going to make use of the data (it's pretty time-consuming). That might not be an issue worth worrying about, though. I think I have changed the structure of the tests pretty significantly since that point. If you're really worried about it, I'll take out the sync.Once stuff.

File: storage/testsuite/long_bench.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (egonelbre) [reply storage/testsuite/long_bench.go]: Initial source download. Large blobs + bad network = terrible experience with a project. I've had to download 1GB repos with 3G. A single blob is never a problem, but it multiple is.

As for modeling https://www.usenix.org/legacy/event/fast09/tech/full_papers/agrawal/agrawal.pdf. However I think it should be completely fine if we only use something like https://play.golang.org/p/gX5Bm3cBQTy... although even the randomization of number of items shouldn't be necessary for benchmarking.

File: storage/testsuite/long_bench.go (hunk 7 above)
Comment (egonelbre) [reply storage/testsuite/long_bench.go]: This slow down would happen only when:
1. you have `--test-bench-long`
2. you have run `-bench PathOperationsInLargeDb/XXX`

Neither happens often, so I wouldn't worry about it and keep code simpler.

File: storage/testsuite/long_bench.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (egonelbre) [reply storage/testsuite/long_bench.go]: _PS: instead of specifying the same parameters for all levels, it could use a folder tree to define the sizes_

[the remark this answers is not in the document]
Comment (thepaul) [reply storage/testsuite/long_bench.go]: That's a very interesting paper! I agree it would be ideal if we had something like that to generate filesystem images, or even if we just had a folder tree with sizes and generated data off of that. However, maybe we can put off implementing that until later, as I don't think it will be worth the time right now with so much to get done. I really suspect that saving hours of developer time at the expense of a few MB of source download is a worthwhile trade, especially when we have hundreds of MB of Go module dependencies that already have to be downloaded.

I'm sympathetic to the argument that small downloads are better, though. Oh, what about if we put the test data in a separate repo somewhere, and you only download it if you care about running the long KVS benchmarks? We could git-submodule it, or just document it somewhere. That wouldn't take much effort to accomplish.

File: storage/boltdb/client_test.go [hunk 10]
```diff
@@ -78,3 +81,56 @@ func TestSuiteShared(t *testing.T) {
...
+func (store *boltLongBenchmarkStore) BulkDelete() error {
+	// do nothing here; everything will be cleaned up later after the test completes. it's not
+	// worth it to wait for BoltDB to remove every key, one by one, and we can't just
+	// os.RemoveAll() the whole test directory at this point because those files are still open
+	// and unremoveable on Windows.
+	return nil
+}
+
+func BenchmarkSuiteLong(b *testing.B) {
+	tempdir, err := ioutil.TempDir("", "storj-bolt")
```
Comment (egonelbre) [inline storage/boltdb/client_test.go]: Since BulkDelete doesn't remove the folder anymore you would need here:
```
defer func(){
    err := os.RemoveAll(tempdir)
    if err != nil { b.Fatal(err) }
}()
```

File: storage/postgreskv/client.go [hunk 11]
```diff
@@ -0,0 +1,296 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+
```
Comment (egonelbre) [inline storage/postgreskv/client.go]: remove this empty line

**Comment by thepaul** [comment]
(last push was a simple squash + rebase)
